### 异常转换(异常封装)
比如：学生上课生病了，此时跟老师上报的异常因该是“不能上课异常”，而不是自己生什么病，给医生上报的是具体的症状异常，
学生生病是自己的异常，自己知道。老师能处理的是不能上课异常，医生处理的病症异常...所以要相应的抛出，保持逻辑清晰
总不能需要老师这个对象应该有个处理学生不能上课异常的逻辑，而不是处理学生生病的逻辑。

```java
	// 添加数据的的时候会抛出 SQLException  
	void addData() throws SQLException {
		......
	}
	,,,
	,,,

	// 逻辑处理，比如添加新用户，当数据库处理失败的时候不应该抛出 SQLException ,而是调用者能理解的，语言清晰的Exception
	void doSth() throws DoFailException {
		try{
			addData();
		}catch(SQLException e){
			.....
			throw new DoFailException();
		}
	}
```

#### 注意事项
由于对象具有多态性,在覆盖父类方法的时候需要保证父类方法能够处理子类对象所抛出的编译检测异常，由此引出两条规则：
* 子类覆盖父类方法时，只能抛出父类的异常或其异常的子类，结合多态可以理解
* 子类方法只能抛出父类方法抛出的异常的子集,同样是多态
上述两条规则针对的仅仅是编译检测异常，对于运行时异常不做规定。
因为代码中父类引用的异常处理是按照父类抛出的异常类做的，所以子类方法抛出的异常父类一定要能识别处理

#### finally
一定会执行的代码
```java
class Test {
	public static void main(String[] args){
		try {
			throw Exception();
		}catch(Exception e) {
			System.out.println(e);

			return; //
		}finally{
			System.out.println("finall");
		}
		System.out.println("xxxx");
	}
}
```
上面的代码中，抛出异常以后有个 return，但是finally 语句块一样会执行，但是后面的代码不会执行了；
但是如果将 return 改为 System.exit(0),则finally语句块中的代码不会执行。
System.exit() 是指退出jvm。

##### try finally 语句块
异常无法直接catch处理，但是资源需要关闭。
```java
try{
	开启资源。。。
	throw Exception();
}finally{
	关闭资源。。。
}
```

在上面的语句块中，抛出异常没catch。就表示这个异常没有处理，应该接着往上抛出。
这种格式的代码经常用于数据库操作之类的逻辑
打开资源和关闭资源应当在同一个方法中处理。在哪里开启的资源，就应该在哪里关闭。
